test/common/http/filter/gzip_filter_test.cc

```diff
@@ -0,0 +1,388 @@
+}
+
+// Transfer-Encoding gzip.
+TEST_F(GzipFilterTest, TransferEncodingGzip) {
+  setUpTest("{}");
+  doRequest({{":method", "get"}, {"accept-encoding", "gzip"}}, true);
+  doResponseNoCompression(
+      {{":method", "get"}, {"content-length", "256"}, {"transfer-encoding", "chunked, deflate"}});
+}
+
+// Content-Encoding: upstream response is already encoded.
+TEST_F(GzipFilterTest, ContentEncodingAlreadyEncoded) {
+  setUpTest("{}");
+  doRequest({{":method", "get"}, {"accept-encoding", "gzip"}}, true);
+  TestHeaderMapImpl response_headers{
+      {":method", "get"}, {"content-length", "256"}, {"content-encoding", "deflate, gzip"}};
+  feedBuffer(256);
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_->encodeHeaders(response_headers, false));
+  EXPECT_TRUE(response_headers.has("content-length"));
+  EXPECT_FALSE(response_headers.has("transfer-encoding"));
+  EXPECT_EQ(FilterDataStatus::Continue, filter_->encodeData(data_, false));
+}
+
+// No compression when upstream response is empty.
+TEST_F(GzipFilterTest, EmptyResponse) {
+  setUpTest("{}");
+  Http::TestHeaderMapImpl headers{{":method", "get"}, {":status", "204"}};
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_->encodeHeaders(headers, true));
+  EXPECT_EQ("", headers.get_("content-length"));
+  EXPECT_EQ("", headers.get_("content-encoding"));
+  EXPECT_EQ(FilterDataStatus::Continue, filter_->encodeData(data_, true));
+}
+
+// Filter should set Vary header value with `accept-encoding`.
+TEST_F(GzipFilterTest, NoVaryHeader) {
+  setUpTest("{}");
+  doRequest({{":method", "get"}, {"accept-encoding", "gzip"}}, true);
+  TestHeaderMapImpl headers{{":method", "get"}, {"content-length", "256"}};
+  feedBuffer(256);
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_->encodeHeaders(headers, false));
+  EXPECT_TRUE(headers.has("vary"));
+  EXPECT_EQ("accept-encoding", headers.get_("vary"));
+}
+
+// Filter should set Vary header value with `accept-encoding` and preserve other values.
+TEST_F(GzipFilterTest, VaryOtherValues) {
+  setUpTest("{}");
+  doRequest({{":method", "get"}, {"accept-encoding", "gzip"}}, true);
+  TestHeaderMapImpl headers{
+      {":method", "get"}, {"content-length", "256"}, {"vary", "user-agent, cookie"}};
+  feedBuffer(256);
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_->encodeHeaders(headers, false));
+  EXPECT_TRUE(headers.has("vary"));
+  EXPECT_EQ("user-agent, cookie, accept-encoding", headers.get_("vary"));
+}
+
+// Vary header should have only one `accept-encoding`value.
+TEST_F(GzipFilterTest, VaryAlreadyHasAcceptEncoding) {
+  setUpTest("{}");
+  doRequest({{":method", "get"}, {"accept-encoding", "gzip"}}, true);
+  TestHeaderMapImpl headers{
+      {":method", "get"}, {"content-length", "256"}, {"vary", "accept-encoding"}};
+  feedBuffer(256);
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_->encodeHeaders(headers, false));
+  EXPECT_TRUE(headers.has("vary"));
+  EXPECT_EQ("accept-encoding", headers.get_("vary"));
+}
+
```

Comment:
can you add some more targeted testing where you directly call isAcceptEncodingAllowed with various values and make sure things look good?

There may be other instances of this.

Reply:
I misinterpreted the RFC. It's fixed now and I also added more tests 👍